# Análisis
---

Para realizar este análisis y en vista de usar una plataforma diferente voy a realizar las queries en Jupyter Notebook y así no tendré que ir realizando cambios entre MySQL Workbench y este Jupyter Notebook.

Dicho esto, las 10 preguntas que voy a intentar resolver son las siguientes:

1. ¿Cuales son los actores con más películas en el DF?
2. ¿Cuales son los actores con menos películas?
3. ¿Cual es la categoría con más películas?
4. ¿Cual es la categoría con menos películas?
5. ¿Cual es la película en la que trabajan más actores?
6. ¿Cual es la película en la que trabajan menos actores?
7. ¿Cual es la película más larga del repertorio?
8. ¿Cual es la película más corta del repertorio?
9. ¿Cuanto fue el tiempo máximo de alquiler de una película?
10. ¿Cuanto fue el tiempo mínimo de alquiler de una película?
  
Vayamos paso a paso...

# Conexión con la BD
---

In [6]:
import mysql.connector as conn
from setting import password_MySQL

# Falta por importar host_MySL y user_MySQL

In [41]:
def querydb(query, limit=5):

    conexion = conn.connect(host='localhost',
                            user='root',
                            passwd=password_MySQL,
                            database='blockbuster')    # conexion al servidor
    cursor = conexion.cursor()
    query += f" limit {limit}"
    cursor.execute(query)
    res = cursor.fetchall()
    print(res)    
    cursor.close()
    conexion.close()

In [35]:
query = 'show databases;'
querydb(query)

[('blockbuster',), ('borrar',), ('information_schema',), ('mysql',), ('performance_schema',), ('sakila',), ('sys',), ('world',)]


In [31]:
query = "SELECT * FROM new_hdd"
querydb(query, 2)

[(1, 1, 1, 6), (2, 23, 1, 2)]


Una vez establecida la conexión empiezo a responder a las preguntas

# Pregunta 1
### ¿Cuales son los actores con más películas en el DF?
---

In [46]:
query = '''
    SELECT 
        actor,
        first_name AS actor_name,
        last_name AS actor_surname,
        COUNT(film) AS films_by_actor
    FROM
        new_hdd
        LEFT JOIN actors ON actors.actor_id = new_hdd.actor
    GROUP BY actor
    ORDER BY films_by_actor DESC
'''
querydb(query, 3)

[(23, 'SANDRA', 'KILMER', 37), (37, 'VAL', 'BOLGER', 35), (13, 'UMA', 'WOOD', 35)]


El top 3 de actores con más películas son:
  
1. Sandra Kilmer con 37 películas.
2. VAL BOLGER con 35 películas.
1. UMA WOOD con 35 películas.

# Pregunta 2
### ¿Cuales son los actores con menos películas?
---

In [48]:
query = '''
    SELECT 
        actor,
        first_name AS actor_name,
        last_name AS actor_surname,
        COUNT(film) AS films_by_actor
    FROM
        new_hdd
        LEFT JOIN actors ON actors.actor_id = new_hdd.actor
    GROUP BY actor
    ORDER BY films_by_actor ASC
'''
querydb(query, 3)

[(39, 'GOLDIE', 'BRODY', 7), (35, 'JUDY', 'DEAN', 15), (31, 'SISSY', 'SOBIESKI', 18)]


El top 3 de actores con menos películas son:
  
1. Goldie Browdy con 7 películas.
2. Judy Dean con 15 películas.
1. Sissy Sobieski con 18 películas.

# Pregunta 3
### ¿Cual es la categoría con más películas?
---

In [51]:
query = '''
    SELECT 
        category,
        categories.name AS category_name,
        COUNT(film) AS films_by_category
    FROM
        new_hdd
        LEFT JOIN categories ON categories.category_id = new_hdd.category
    GROUP BY category
    ORDER BY films_by_category DESC
'''
querydb(query, 3)

[(15, 'Sports', 79), (1, 'Action', 77), (2, 'Animation', 75)]


Las categorías con más películas son:
  
  1. Sports con 79 películas.
  2. Action con 77 películas.
  3. Animation con 75 películas.

# Pregunta 4
### ¿Cual es la categoría con menos películas?
---

In [57]:
query = '''
    SELECT 
        category,
        categories.name AS category_name,
        COUNT(film) AS films_by_category
    FROM
        new_hdd
        LEFT JOIN categories ON categories.category_id = new_hdd.category
    GROUP BY category
    ORDER BY films_by_category ASC
'''
querydb(query, 3)

[(10, 'Games', 38), (5, 'Comedy', 46), (16, 'Travel', 52)]


Las categorías con menos películas son:
  
  1. Games con 38 películas.
  2. Comedy con 46 películas.
  3. Travel con 52 películas.

# Pregunta 5
### ¿Cual es la película en la que trabajan más actores?
---


In [56]:
query = '''
    SELECT 
        film.title,
        COUNT(actor) AS actors_by_film
    FROM
        new_hdd
        LEFT JOIN film ON film.film_id = new_hdd.film
    GROUP BY film
    ORDER BY actors_by_film DESC
'''
querydb(query, 3)

[('BOONDOCK BALLROOM', 6), ('FLASH WARS', 5), ('OKLAHOMA JUMANJI', 5)]


Las peliculas en las que trabajan más actores son:
  
  1. Boondock ballroom con 6 actores.
  2. Flash wars con 5 actores.
  3. Oklahoma Jumanji con 5 actores.

# Pregunta 6
### ¿Cual es la película en la que trabajan menos actores?
---


In [55]:
query = '''
    SELECT 
        film.title,
        COUNT(actor) AS actors_by_film
    FROM
        new_hdd
        LEFT JOIN film ON film.film_id = new_hdd.film
    GROUP BY film
    ORDER BY actors_by_film ASC
'''
querydb(query, 3)

[('EXORCIST STING', 1), ('COLOR PHILADELPHIA', 1), ('GLEAMING JAWBREAKER', 1)]


Las peliculas en las que trabajan menos actores son:
  
  1. Exorcist sting con 1 actores.
  2. Color Philadelphia con 1 actores.
  3. Gleaming Jawbreaker con 1 actores.

# Pregunta 7
### ¿Cual es la película más larga del repertorio?
---


In [58]:
query = '''
    SELECT 
        film.title,
        film.length AS length
    FROM
        film
    ORDER BY length DESC
'''
querydb(query, 3)

[('DARN FORRESTER', 185), ('CHICAGO NORTH', 185), ('CONTROL ANTHEM', 185)]


Las peliculas más largas del repertorio son:
  
  1. Darn Forrester con 185 minutos.
  2. Chicago North con 185 minutos.
  3. Control Anthem con 185 minutos.

# Pregunta 8
### ¿Cual es la película más corta del repertorio?
---

In [59]:
query = '''
    SELECT 
        film.title,
        film.length AS length
    FROM
        film
    ORDER BY length ASC
'''
querydb(query, 3)

[('IRON MOON', 46), ('KWAI HOMEWARD', 46), ('ALIEN CENTER', 46)]


Las peliculas más largas del repertorio son:
  
  1. Iron Moon con 46 minutos.
  2. Kawai Homeward con 46 minutos.
  3. Alien Center con 46 minutos.

# Pregunta 9
### ¿Cuanto fue el tiempo máximo de alquiler de una película?
---


In [60]:
query = '''
    SELECT 
        film.title,
        film.rental_duration AS ren_duration
    FROM
        film
    ORDER BY ren_duration DESC
'''
querydb(query, 3)

[('ARGONAUTS TOWN', 7), ('ADAPTATION HOLES', 7), ('ANONYMOUS HUMAN', 7)]


Las peliculas que han tenido una mayor duración de alquiler son:
  
  1. Argonauts Town con 7 días.
  2. Adaptation Holes con 7 días.
  3. Anonymous Human con 7 días.

# Pregunta 10
### ¿Cuanto fue el tiempo mínimo de alquiler de una película?
---


In [61]:
query = '''
    SELECT 
        film.title,
        film.rental_duration AS ren_duration
    FROM
        film
    ORDER BY ren_duration ASC
'''
querydb(query, 3)

[('ALABAMA DEVIL', 3), ('ACE GOLDFINGER', 3), ('AGENT TRUMAN', 3)]


Las peliculas que han tenido una menor duración de alquiler son:
  
  1. Alabama Devil con 3 días.
  2. Ace Goldfinger con 3 días.
  3. Agent Truman con 3 días.